### PR TITLE
Test assembler opcodes

### PR DESCRIPTION
**Pull request by AdamuKaapan**
This adds the DC command, LEA command (#13), SIMHALT command (#49), and END command. Also fixed a couple assembler bugs that I discovered while implementing LEA.

### CODE REVIEW

File: src/easier68k/core/opcodes/dc.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+    @staticmethod
+    def is_valid(command: str, parameters: str) -> (bool, list):
+        """
+        Tests whether the given command is valid
+
+        >>> DC.is_valid('DC.B', '\\'Hello, world!\\'')[0]
+        True
+
+        >>> DC.is_valid('DC.B', '\\'Hello\\'\\' world!\\'')[0]
```
Comment (Chris-Johnston) [inline src/easier68k/core/opcodes/dc.py]: This is where I would give up on doctests and start using pytest. This is good, but maybe in future it would be more readable to do this in a pytest. Escaping looks to be kinda tricky.

File: src/easier68k/core/opcodes/dc.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+                    is_string = False
+                    quote_delim = None
+                    current_param = ''
+                    continue
+
+                if quote_delim and c == quote_delim:  # Breaking out of the quote
+                    quote_delim = None
+                    continue
+
+                if c == '\'':
```
Comment (Chris-Johnston) [inline src/easier68k/core/opcodes/dc.py]: Again, would be good to use constants here.

File: src/easier68k/core/opcodes/dc.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+                    if not quote_delim:  # This is the start of a quote
+                        quote_delim = '\''
+                        assert not current_param.strip(), "Can't have two quotes in a row in one parameter"
+                        current_param = ''
+                        is_string = True
+                        continue
+
+                if c == '\"':
+                    if not quote_delim:  # This is the start of a quote
+                        quote_delim = '\"'
```
Comment (Chris-Johnston) [inline src/easier68k/core/opcodes/dc.py]: Are you certain 68k supports `"` and `'` for their strings? I thought this was only `"`.

File: src/easier68k/core/opcodes/dc.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+                    continue
+
+                if quote_delim and c == quote_delim:  # Breaking out of the quote
+                    quote_delim = None
+                    continue
+
+                if c == '\'':
+                    if not quote_delim:  # This is the start of a quote
+                        quote_delim = '\''
+                        assert not current_param.strip(), "Can't have two quotes in a row in one parameter"
```
Comment (Chris-Johnston) [inline src/easier68k/core/opcodes/dc.py]: I would re-word this assert, I'm unsure what it is trying to tell me. Maybe instead say "Cannot have an empty string?"

File: src/easier68k/core/opcodes/dc.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+                    if not quote_delim:  # This is the start of a quote
+                        quote_delim = '\"'
+                        assert not current_param.strip(), "Can't have two quotes in a row in one parameter"
+                        current_param = ''
+                        is_string = True
+                        continue
+
+                current_param += c
+
+            assert not quote_delim, 'Unclosed quotes'
```
Comment (Chris-Johnston) [inline src/easier68k/core/opcodes/dc.py]: I would reword the assertion message.

File: src/easier68k/core/opcodes/dc.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+        >>> DC.get_word_length('DC.W', '\\'Hai!\\'')[0]
+        2
+
+        >>> DC.get_word_length('DC.L', '\\'Hai!\\'')[0]
+        4
+
+        :param command: The command itself (e.g. 'MOVE.B', 'LEA', etc.)
+        :param parameters:  The parameters after the command (such as the source and destination of a move)
+        :return: The length of the command in memory (in words) and a list of issues/warnings encountered during assembly
```
Comment (Chris-Johnston) [inline src/easier68k/core/opcodes/dc.py]: Why measure length in words and not length in bytes? I'm fairly certain elsewhere measures in bytes.
This would also get around the issues of adding 1/2 for byte values.

File: src/easier68k/core/opcodes/dc.py [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+    def execute(self, simulator: M68K):
+        """
+        Executes this command in a simulator
+        :param simulator: The simulator to execute the command on
+        :return: Nothing
+        """
+        pass
+
+    def __str__(self):
+        return "DC command: Size {}, items: {}".format(self.size, self.values)
```
Comment (Chris-Johnston) [inline src/easier68k/core/opcodes/dc.py]: I wonder if we could make opcodes more OOP, since we'll have to be making so many of them.
Instead of defining `__str__` for each opcode, could have a value in the class that specifies what the opcode name is, and a list of its parameters.

File: src/easier68k/core/opcodes/lea.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+def command_matches(command: str) -> bool:
+    """
+    Checks whether a command string is an instance of this command type
+    :param command: The command string to check (e.g. 'MOVE.B', 'LEA', etc.)
+    :return: Whether the string is an instance of this command type
+    """
+    return opcode_util.command_matches(command, 'LEA')
+
+
+class_name = 'Lea'
```
Comment (Chris-Johnston) [inline src/easier68k/core/opcodes/lea.py]: don't think this makes sense to have in the module level

that, and I don't see if/where it is used.

File: src/easier68k/core/opcodes/lea.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        :param command: The command itself (e.g. 'MOVE.B', 'LEA', etc.)
+        :param parameters: The parameters after the command (such as the source and destination of a move)
+        :return: Whether the given command is valid and a list of issues/warnings encountered
+        """
+        issues = []
+        try:
+            assert opcode_util.check_valid_command(command, 'LEA', can_take_size=False), 'Command invalid'
+
+            # Split the parameters into EA modes
+            params = parameters.split(',')
```
Comment (Chris-Johnston) [inline src/easier68k/core/opcodes/lea.py]: Why not use that utility function?

Or, was that in the PR I'm working on that hasn't been merged yet?

File: src/easier68k/core/opcodes/dc.py [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+                continue
+
+            if c == '\'':
+                if not quote_delim:  # This is the start of a quote
+                    quote_delim = '\''
+                    current_param = ''
+                    is_string = True
+                    continue
+
+            if c == '\"':
```
[the remark this answers is not in the document]
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/dc.py]: I don't think so personally (especially since this is being changed drastically anyways so now we'll only have one type of quote to look for), I feel like for how little this is used it's fine as just being hard-coded (but if you think it's really a big deal I can change it)

File: src/easier68k/core/opcodes/dc.py [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+        >>> test0.size
+        'B'
+        >>> test0.values
+        ['$0A', '$0B']
+
+        >>> test1 = DC.from_str("DC.B", "\\'Hai!\\'")[0]
+        >>> test1.size
+        'B'
+        >>> test1.values
+        ['$48', '$61', '$69', '$21']
```
[the remark this answers is not in the document]
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/dc.py]: Good question, I think I can convert it to numbers later on; I think it happened because I was doing some weird stuff to make sure the zero-padding got saved right, but looking at how it's set up right now I can probably make the final step convert to numbers instead of strings.

File: src/easier68k/core/opcodes/dc.py [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+
+        if current_param:
+            if is_string:
+                temp_params = []
+                for c in current_param:
+                    temp_params.append('$' + format(ord(c), "x"))
+
+                # Right pad the string with zeroes
+                if size == 'L':
+                    while len(temp_params) % 4 != 0:
```
[the remark this answers is not in the document]
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/dc.py]: Probably, yeah, that's a good idea: found a nice simple way to do that with a quick search.

File: src/easier68k/core/opcodes/dc.py (hunk 1 above)
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/dc.py]: Escaping is super weird, yeah: I don't think that'll come up anywhere else though since DC is (afaik) the only place that uses string literals (?).

File: src/easier68k/core/opcodes/dc.py (hunk 6 above)
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/dc.py]: I had some reason for it, I think it's mainly because instructions never come out to byte-level lengths (they're always an even word length): I add half here because it gets rounded up in the end. It would work either way, so I'm up for changing it, it just has to remain consistent for all opcodes.

File: src/easier68k/core/opcodes/dc.py (hunk 3 above)
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/dc.py]: After a quick test, can confirm that it only assembles using `'` for string literals, and it then uses `''` to define an apostrophe inside the literal, so I'll need to go ahead and redo that.

File: src/easier68k/core/opcodes/dc.py (hunk 4 above)
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/dc.py]: It's mainly trying to say that I can't do, for example `'Hello' ' world'`, defining two string literals in a row without them being separated by a comma. I might change it to "Can't have two string literals in a row without being separated by a comma", which while still kinda weird (I can't think of a good way to put this error) should do a bit better.

File: src/easier68k/core/opcodes/dc.py (hunk 7 above)
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/dc.py]: There's already the `class_name` parameter which we can use and then do some introspection to print it out and/or make a method like `get_str` in Opcode which takes the class name and a list of variables and pretty-prints them kinda like this.

On the note of making them more OOP, I did have the idea of making a couple base classes like "two-parameter opcode", which has two assembly parameters and you just define legal types for both of them and then override execute and assemble, while that base class automates some of the parsing.

File: src/easier68k/core/opcodes/lea.py (hunk 8 above)
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/lea.py]: It's used in the assembler, because we need to use introspection to get the class itself (which is why it's at the module level): I can probably make it just look for the first class in the module and use that instead.

File: src/easier68k/core/opcodes/lea.py (hunk 9 above)
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/lea.py]: That hasn't been merged yet/this isn't rebased yet where that exists (one way or another it's not here yet but will be soon)

File: src/easier68k/core/opcodes/dc.py (hunk 4 above)
Comment (LinuxKid21) [reply src/easier68k/core/opcodes/dc.py]: Why not something like "expected comma between two string literals" ?

Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/dc.py]: ..... I like that much better, that was obvious, let's go with that

File: src/easier68k/core/opcodes/dc.py (hunk 6 above)
Comment (Chris-Johnston) [reply src/easier68k/core/opcodes/dc.py]: Could multiply the result to return bytes so that it can be used elsewhere.

File: src/easier68k/core/opcodes/lea.py (hunk 8 above)
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/lea.py]: This will be handled in another PR (looks like #79 )

File: src/easier68k/core/opcodes/dc.py (hunk 7 above)
Comment (AdamuKaapan) [reply src/easier68k/core/opcodes/dc.py]: This will be handled in another PR
